File: levelledmobs-plugin-bukkit/src/main/java/me/lokka30/levelledmobs/bukkit/listener/ListenerHandler.java
package me.lokka30.levelledmobs.bukkit.listener;

import java.util.Set;
import me.lokka30.levelledmobs.bukkit.logic.function.process.condition.impl.ConditionParseListener;
import me.lokka30.levelledmobs.bukkit.util.Log;
import org.jetbrains.annotations.NotNull;

public final class ListenerHandler {

    /* vars */

    /**
     * These listeners must be registered earlier.
     */
    private final Set<ListenerWrapper> primaryListeners = Set.of(
        new ActionParseListener(),
        new ConditionParseListener(),
        new LevellingStrategyRequestListener(),
        new TestListener()
    );

    /**
     * These listeners must be registered later.
     */
    private final Set<ListenerWrapper> secondaryListeners = Set.of(
        new EntitySpawnListener(),
        new PlayerJoinListener()
    );

    /* methods */

    public boolean loadPrimary() {
        Log.inf("Registering primary listeners");
        for(var listener : getPrimaryListeners())
            if(!listener.register())
                return false;
        return true;
    }

    public boolean loadSecondary() {
        Log.inf("Registering secondary listeners");
        for(var listener : getSecondaryListeners())
            if(!listener.register())
                return false;
        return true;
    }

    /* getters and setters */

    @NotNull
    public Set<ListenerWrapper> getPrimaryListeners() { return primaryListeners; }

    @NotNull
    public Set<ListenerWrapper> getSecondaryListeners() { return secondaryListeners; }

}

File: levelledmobs-plugin-bukkit/src/main/kotlin/io/github/arcaneplugins/levelledmobs/bukkit/listener/ListenerHandler.kt
package io.github.arcaneplugins.levelledmobs.bukkit.listener

import io.github.arcaneplugins.levelledmobs.bukkit.listener.impl.*
import io.github.arcaneplugins.levelledmobs.bukkit.util.Log.inf

object ListenerHandler {
    private val primaryListeners = mutableSetOf(
        ActionParseListener(),
        ConditionParseListener(),
        LevellingStrategyRequestListener(),
        TestListener()
    )

    private val secondaryListeners = mutableSetOf(
        EntityBreedListener(),
        EntityDamageByEntityListener(),
        EntityDamageListener(),
        EntityDeathListener(),
        EntityExplodeListener(),
        EntityRegainHealthListener(),
        EntitySpawnListener(),
        EntityTransformListener(),
        PlayerDeathListener(),
        PlayerJoinListener(),
        PlayerQuitListener(),
        PlayerTeleportListener()
    )

    fun loadPrimary(){
        inf("Registering primary listeners")
        primaryListeners.forEach(ListenerWrapper::register)
    }

    fun loadSecondary(){
        inf("Registering secondary listeners")
        secondaryListeners.forEach(ListenerWrapper::register)
    }
}